Cleanup directionality code

rniwa opened the pull request:
#### 2622f839cafe16caaca28a445723699c3e7607b3
<pre>
Cleanup directionality code
<a href="https://bugs.webkit.org/show_bug.cgi?id=244076">https://bugs.webkit.org/show_bug.cgi?id=244076</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::defaultWritingDirection): Made this function return std::optional&lt;UCharDirection&gt;
instead of taking an out argument.
* Source/WTF/wtf/text/StringImpl.h:

* Source/WTF/wtf/text/WTFString.h:
(WTF::String::defaultWritingDirection const): Ditto.

* Source/WebCore/html/HTMLElement.cpp:
(WebCore::isLTROrRTLIgnoringCase): Deleted.
(WebCore::HTMLElement::collectPresentationalHintsForAttribute):
(WebCore::HTMLElement::computeDirectionality const):
(WebCore::HTMLElement::directionalityIfDirIsAuto const): Renamed from directionalityIfhasDirAutoAttribute.
This function now returns TextDirectionAutoResult instead of taking a boolean out argument.
(WebCore::HTMLElement::directionality const): Now returns TextDirectionWithStrongDirectionalityNode
instead of taking pointer to Node* as an out argument.
(WebCore::HTMLElement::adjustDirectionalityIfNeededAfterChildAttributeChanged):
(WebCore::HTMLElement::calculateAndAdjustDirectionality):

* Source/WebCore/html/HTMLElement.h:
(WebCore::HTMLElement::TextDirectionWithStrongDirectionalityNode):

* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::directionForFormData const):

* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::setValue):

* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::matchAllRules):

Canonical link: <a href="https://commits.webkit.org/253589@main">https://commits.webkit.org/253589@main</a>
</pre>













<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1cc502ca5ac40b1feb251e67fada7c23fb2671f7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86445 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30365 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17413 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95292 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149001 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/90433 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/28765 "Built successfully") | [✅ ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78599 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90535 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92061 "Passed tests") | [✅ ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/23345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/73415 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23397 "Passed tests") | 
| | [✅ ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/78354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/78742 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66401 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/78382 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/26678 "Built successfully") | [✅ ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/12592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72017 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26592 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13604 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25742 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28271 "Built successfully") | [✅ ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/73415 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/74798 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/986 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28211 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/32887 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16539 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -803,62 +797,50 @@ TextDirection HTMLElement::computeDirectionality() const
-                *strongDirectionalityTextNode = hasStrongDirectionality ? textControl : nullptr;
-            return (textDirection == U_LEFT_TO_RIGHT) ? TextDirection::LTR : TextDirection::RTL;            
+            auto direction = textControl->value().defaultWritingDirection();
+            if (direction == std::nullopt)
```
We typically use the `!` operator in a case like this.

rniwa replied:
Fixed.

darinadler commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -803,62 +797,50 @@ TextDirection HTMLElement::computeDirectionality() const
-                    *strongDirectionalityTextNode = node.get();
-                return (textDirection == U_LEFT_TO_RIGHT) ? TextDirection::LTR : TextDirection::RTL;
-            }
+            auto direction = node->textContent(true).defaultWritingDirection();
```
Put this inside the `if`?

rniwa replied:
Done.

darinadler commented on Source/WebCore/html/HTMLElement.h:
```diff
@@ -90,7 +90,12 @@ class HTMLElement : public StyledElement {
+        TextDirection direction;
+        bool isAuto { false };
+    };
+    TextDirectionAutoResult directionalityAndIsAuto() const;
```
I find this function difficult to understand. Despite its name it does not simply return directionality and also a boolean to indicate whether auto applies; the auto state affects the directionality. Separately, I think `isAuto` is subtle and ambiguous; what is auto and what does auto mean? I think it might need a different name and return type to clarify what’s going on. The old name seems slightly clearer. Maybe just revert to that name and return an `optional`? We would have to use `value_or` a little bit but I think it might be worth it for slightly better clarity.

rniwa replied:
I'd rename it to directionalityIfDirIsAuto and make it return std::optional<TextDirection>.

darinadler commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -803,62 +797,50 @@ TextDirection HTMLElement::computeDirectionality() const
-            }
+            auto direction = node->textContent(true).defaultWritingDirection();
+            if (direction)
+                return { *direction == U_LEFT_TO_RIGHT ? TextDirection::LTR : TextDirection::RTL, node.get() };
```
Why is `get()` needed here?

rniwa replied:
That was unnecessary. Removed.

rniwa committed:
Cleanup directionality code
https://bugs.webkit.org/show_bug.cgi?id=244076

Reviewed by Darin Adler.

* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::defaultWritingDirection): Made this function return std::optional<UCharDirection>
instead of taking an out argument.
* Source/WTF/wtf/text/StringImpl.h:

* Source/WTF/wtf/text/WTFString.h:
(WTF::String::defaultWritingDirection const): Ditto.

* Source/WebCore/html/HTMLElement.cpp:
(WebCore::isLTROrRTLIgnoringCase): Deleted.
(WebCore::HTMLElement::collectPresentationalHintsForAttribute):
(WebCore::HTMLElement::computeDirectionality const):
(WebCore::HTMLElement::directionalityIfDirIsAuto const): Renamed from directionalityIfhasDirAutoAttribute.
This function now returns TextDirectionAutoResult instead of taking a boolean out argument.
(WebCore::HTMLElement::directionality const): Now returns TextDirectionWithStrongDirectionalityNode
instead of taking pointer to Node* as an out argument.
(WebCore::HTMLElement::adjustDirectionalityIfNeededAfterChildAttributeChanged):
(WebCore::HTMLElement::calculateAndAdjustDirectionality):

* Source/WebCore/html/HTMLElement.h:
(WebCore::HTMLElement::TextDirectionWithStrongDirectionalityNode):

* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::directionForFormData const):

* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::setValue):

* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::matchAllRules):

Canonical link: https://commits.webkit.org/253589@main
